How do we classify the data which have more thn 2 categories or have 2 categories which are not binary in Nature using 
Decision tree 

Concept of Entropy and info gain help to build a Decision tree model

How to build a decision tree in Python

If a data which you want to classify doesnt have a target variable with 2 categories or have to categories but this
arent in binary in nature, we can not use logistic regression. In such case, we need to depend ontwo CART(Classification
and Regression Techniques) model:

a) Decision Tree
b) Random Forest

Both of them are know as CART model coz, we can use them for classification as well as for Regression

A decision tree, consist of Nodes and Branches a node is a input variable which split the data.
Nodes can further be classified as primary node and secondary node.

A primary node is the inout which slit the data for the first time and have max impact on targetvariable with highest info gain

A secondary nodes are then arranged in the sequesce in which they impact the target variable.

A decision tree will calculate the information gain for each input variable to find a primary and secondary nodes using the concept of entropy and info gain.

NOTE : Please refer word file for entropy and info gain

In [1]:
import pandas as pd
from sklearn.model_selection import train_test_split

In [2]:
#Step 1:Import data
df = pd.read_csv(r"E:/Dataset/winequality-red.csv")
df

fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
0               7.4             0.700         0.00             1.9      0.076   
1               7.8             0.880         0.00             2.6      0.098   
2               7.8             0.760         0.04             2.3      0.092   
3              11.2             0.280         0.56             1.9      0.075   
4               7.4             0.700         0.00             1.9      0.076   
...             ...               ...          ...             ...        ...   
1594            6.2             0.600         0.08             2.0      0.090   
1595            5.9             0.550         0.10             2.2      0.062   
1596            6.3             0.510         0.13             2.3      0.076   
1597            5.9             0.645         0.12             2.0      0.075   
1598            6.0             0.310         0.47             3.6      0.067   

      free sulfur dioxide  total sulfur dioxide  density    pH  sulphates  \
0                    11.0                  34.0  0.99780  3.51       0.56   
1                    25.0                  67.0  0.99680  3.20       0.68   
2                    15.0                  54.0  0.99700  3.26       0.65   
3                    17.0                  60.0  0.99800  3.16       0.58   
4                    11.0                  34.0  0.99780  3.51       0.56   
...                   ...                   ...      ...   ...        ...   
1594                 32.0                  44.0  0.99490  3.45       0.58   
1595                 39.0                  51.0  0.99512  3.52       0.76   
1596                 29.0                  40.0  0.99574  3.42       0.75   
1597                 32.0                  44.0  0.99547  3.57       0.71   
1598                 18.0                  42.0  0.99549  3.39       0.66   

      alcohol  quality  
0         9.4        5  
1         9.8        5  
2         9.8        5  
3         9.8        6  
4         9.4        5  
...       ...      ...  
1594     10.5        5  
1595     11.2        6  
1596     11.0        6  
1597     10.2        5  
1598     11.0        6  

[1599 rows x 12 columns]

# Using the above data make a decision tree to classify any sample of a new wine into quality categories also find which of the input variable gives max information about the quality of wine?

In decision tree target var should always be categorical in nature. Here quality is taget var 


In [3]:
df.dtypes

fixed acidity           float64
volatile acidity        float64
citric acid             float64
residual sugar          float64
chlorides               float64
free sulfur dioxide     float64
total sulfur dioxide    float64
density                 float64
pH                      float64
sulphates               float64
alcohol                 float64
quality                   int64
dtype: object

The quality is int so we have to convert it into categorical

In [4]:
df['quality'] = df['quality'].astype("category")

In [5]:
df.dtypes

fixed acidity            float64
volatile acidity         float64
citric acid              float64
residual sugar           float64
chlorides                float64
free sulfur dioxide      float64
total sulfur dioxide     float64
density                  float64
pH                       float64
sulphates                float64
alcohol                  float64
quality                 category
dtype: object

In [6]:
# Also check if any col has missing value
df.isnull().sum()

fixed acidity           0
volatile acidity        0
citric acid             0
residual sugar          0
chlorides               0
free sulfur dioxide     0
total sulfur dioxide    0
density                 0
pH                      0
sulphates               0
alcohol                 0
quality                 0
dtype: int64

In [7]:
#Step 2: lets build the decision tree

# Find your input and target variable 
Y = df[['quality']]
X = df.drop(columns = ['quality'])

# Step 3: Split the data
X_train, X_test, Y_train, Y_test = train_test_split(X,Y, train_size = 0.8, random_state=1234)
 
len(X_train), len(X_test), len(Y_train), len(Y_test)

(1279, 320, 1279, 320)

In [8]:
# Step 4: Creating the model using the training data set
from sklearn.tree import DecisionTreeClassifier
dtree = DecisionTreeClassifier()

model = dtree.fit(X_train, Y_train)

model

DecisionTreeClassifier()

In [9]:
# STEP 5: Using this model Predict the X_test Case

Y_test['Pred_Quality'] = model.predict(X_test)

Y_test

quality  Pred_Quality
688        5             5
961        5             5
726        6             5
537        6             5
1544       7             6
...      ...           ...
351        6             6
415        5             5
564        6             6
1124       4             7
147        5             6

[320 rows x 2 columns]

In [10]:
# STEP 6: To print the classification report

from sklearn.metrics import classification_report

print(classification_report(Y_test['quality'], Y_test['Pred_Quality']))

              precision    recall  f1-score   support

           3       0.00      0.00      0.00         2
           4       0.00      0.00      0.00         8
           5       0.77      0.72      0.74       138
           6       0.65      0.66      0.65       131
           7       0.52      0.69      0.59        39
           8       0.00      0.00      0.00         2

    accuracy                           0.66       320
   macro avg       0.32      0.34      0.33       320
weighted avg       0.66      0.66      0.66       320



C:\Users\dell\anaconda3\lib\site-packages\sklearn\metrics\_classification.py:1318: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
C:\Users\dell\anaconda3\lib\site-packages\sklearn\metrics\_classification.py:1318: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
C:\Users\dell\anaconda3\lib\site-packages\sklearn\metrics\_classification.py:1318: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


In [11]:
# Lets check the number of observations in each category

print("Count of 3 in Y_train", sum(Y_train['quality'] == 3))
print("Count of 4 in Y_train", sum(Y_train['quality'] == 4))
print("Count of 5 in Y_train", sum(Y_train['quality'] == 5))
print("Count of 6 in Y_train", sum(Y_train['quality'] == 6))
print("Count of 7 in Y_train", sum(Y_train['quality'] == 7))
print("Count of 8 in Y_train", sum(Y_train['quality'] == 8))

Count of 3 in Y_train 8
Count of 4 in Y_train 45
Count of 5 in Y_train 543
Count of 6 in Y_train 507
Count of 7 in Y_train 160
Count of 8 in Y_train 16


Count of 3 in Y_train 8
Count of 4 in Y_train 45
Count of 5 in Y_train 543
Count of 6 in Y_train 507
Count of 7 in Y_train 160
Count of 8 in Y_train 16
Thus, we must use SMOTE to balance this data and rebuild the model again to get a better model to predict the wine quality.


Apply SMOTE on X_train and Y_train and re-build the model to find the recall value of each category of wine and the overall accuracy of the model.

In [12]:
# To find the information gain for each input variable, we can use a 
# command feature_importances_ on a Model object

dtree.feature_importances_

array([0.06845364, 0.0908174 , 0.05227318, 0.0632756 , 0.06266574,
       0.07057434, 0.1222004 , 0.07516849, 0.0915633 , 0.12164003,
       0.18136789])

Although the overall accuracy of the model is 67%, but the recall value of each category differ a lot. Infact, the model does not know how to classify the wines in category 3, 4, and 8 as the recall value for each of these category is 0. This bcoz the data have imbalance in it.

In [13]:
# Lets create this as pandas data frame

IG_DF = pd.DataFrame({"Input Variable" : X_train.columns,
                     "IG Value" : dtree.feature_importances_})

IG_DF.sort_values(by = "IG Value")

Input Variable  IG Value
2            citric acid  0.052273
4              chlorides  0.062666
3         residual sugar  0.063276
0          fixed acidity  0.068454
5    free sulfur dioxide  0.070574
7                density  0.075168
1       volatile acidity  0.090817
8                     pH  0.091563
9              sulphates  0.121640
6   total sulfur dioxide  0.122200
10               alcohol  0.181368